### PR TITLE
Infoblox BloxOne Threat Defense

### PR DESCRIPTION
**Pull request by esharf**
## Status
- [x] Ready

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-4728

## Description
added a new integration **Infoblox BloxOne Threat Defense**

## Minimum version of Cortex XSOAR
- [x] 6.8.0

## Does it break backward compatibility?
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+                              limit: int = 50, offset: Optional[int] = None) -> List[Dict]:
+        url_suffix = '/api/tdlad/v1/lookalike_domains'
+        filter_params = {key: val for key, val in [('_limit', limit), ('_offset', offset)] if val}
+
+        if user_filter:
+            _filter = user_filter
+        elif target_domain:
+            _filter = f'target_domain=="{target_domain}"'
+        else: # detected_at != None
+            _filter = f'detected_at>="{detected_at}"'
```
Comment (DeanArbel) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.py:31]: Can't the filter be used with more than one filter value?
If not, please mention it in the description.

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    return CommandResults(
+        outputs_prefix='BloxOneTD',
+        outputs={'DossierSource': sources},
+    )
+
+
+def validate_and_format_lookalike_domain_list_args(args: Dict) -> Dict:
+    match len(list(filter(bool, [args.get('filter'), args.get('target_domain'), args.get('detected_at')]))):
+        case 0:
+            raise ValueError("You must spesify one of the following arguments 'target_domain', 'detected_at' or 'filter'.")
```
Comment (DeanArbel) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.py:82]:
```suggestion
            raise ValueError("Please provide one of the following arguments 'target_domain', 'detected_at' or 'filter'.")
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        case 1:
+            pass
+        case _:
+            raise ValueError("You can spesify one of the following arguments 'target_domain', 'detected_at' or 'filter' not more.")
```
Comment (DeanArbel) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.py:86]:
```suggestion
            raise ValueError("Exactly one of the following arguments must be provided 'target_domain', 'detected_at' or 'filter'. No more than one is argument is accepted")
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    args['timeout'] = timeout_in_seconds - next_run_in_seconds
+    scheduled_command = ScheduledCommand(
+        command='bloxone-td-dossier-lookup-get',
+        next_run_in_seconds=next_run_in_seconds,
+        timeout_in_seconds=timeout_in_seconds,
+        args=args
+    )
+
+    return CommandResults(
+        readable_output='Job is still running, it may take a little while...',
```
Comment (DeanArbel) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.py:136]: If you remove this part, there won't be any print after the first execution. Please consider avoiding further prints after the first one.

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/test_data/baseintegration-dummy.json [hunk 5]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (DeanArbel) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/test_data/baseintegration-dummy.json:1]: ???

**Review by DeanArbel — CHANGES_REQUESTED** [review]
Mostly looks ok, please note the images need to be updated.
Please add tests.

**Comment by esharf** [comment]
i have fixed the comments.
still need to add Tests and Docs

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+  section: Connect
+- display: Trust any certificate (not secure)
+  name: insecure
+  required: false
+  type: 8
+- display: Use system proxy settings
+  name: proxy
+  required: false
+  type: 8
+description: 'BloxOne Threat Defense is a hybrid cybersecurity solution that leverages DNS as the first line of defense to detect and block cyber threats.'
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml:20]:
```suggestion
description: BloxOne Threat Defense is a hybrid cybersecurity solution that leverages DNS as the first line of defense to detect and block cyber threats.
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+      - hash
+      - email
+    - default: false
+      description: 'The indicator to search on.'
+      isArray: false
+      name: value
+      required: true
+      secret: false
+    - default: false
+      description: 'The sources to query. Multiple sources can be specified, if no source is specified, the call will search on all available sources. (you can see the list of the available sources by runing **bloxone-td-dossier-source-list**).'
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml:46]:
```suggestion
      description: 'The sources to query. Multiple sources can be specified. If no source is specified, the call will search on all available sources. (You can see the list of the available sources by running bloxone-td-dossier-source-list).'
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    description: 'The Dossier Lookup API returns detailed information on the specified indicator from the requested sources.'
+    execution: false
+    polling: true
+    name: bloxone-td-dossier-lookup-get
+    outputs:
+    - contextPath: BloxOneTD.DossierLookup.source
+      description: 'The Dossier source.'
+      type: String
+    - contextPath: BloxOneTD.DossierLookup.target
+      description: 'The targeted indicator.'
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml:77]:
```suggestion
      description: The targeted indicator.
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+      type: String
+    - contextPath: BloxOneTD.DossierLookup.task_id
+      description: 'The Dossier task id.'
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml:80]:
```suggestion
      description: The Dossier task ID.
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+      type: String
+    - contextPath: BloxOneTD.DossierLookup.type
+      description: 'The indicator type.'
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml:83]:
```suggestion
      description: The indicator type.
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+      type: String
+  - name: bloxone-td-dossier-source-list
+    description: 'Get available Dossier sources.'
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml:86]:
```suggestion
    description: Get available Dossier sources.
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    outputs:
+    - contextPath: BloxOneTD.DossierSource
+      description: 'Available Dossier sources.'
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml:89]:
```suggestion
      description: Available Dossier sources.
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+      type: String
+  - arguments:
+    - default: false
+      description: 'The free query filter argument.'
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml:93]:
```suggestion
      description: The free query filter argument.
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+      required: false
+      secret: false
+    - default: false
+      description: Filter by detected at greater then or equal to. you can use ISO format (e.g. '2023-02-14T00:11:22Z') or use a relative time (e.g. "3 days").
+      isArray: false
+      name: detected_at
+      required: false
+      secret: false
+    - default: false
+      description: 'Maximum number of results to return from the query. Default is 50.'
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml:111]:
```suggestion
      description: Maximum number of results to return from the query. 
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+      isArray: false
+      name: limit
+      required: false
+      secret: false
+      defaultValue: 50
+    - default: false
+      description: 'Return results starting at this offset. should be int - by default is 0'
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml:118]:
```suggestion
      description: Return results starting at this offset. Should be an integer. Default is 0.
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+      isArray: false
+      name: offset
+      required: false
+      secret: false
+    deprecated: false
+    description: 'Get Lookalike Domain Lists.'
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml:124]:
```suggestion
    description: 'Get Lookalike domain lists.'
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+    execution: false
+    polling: true
+    name: bloxone-td-lookalike-domain-list
+    outputs:
+    - contextPath: BloxOneTD.LookalikeDomain.detected_at
+      description: 'The date of the lookalike detection.'
+      type: Date
+    - contextPath: BloxOneTD.LookalikeDomain.lookalike_domain
+      description: 'The lookalike domain.'
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml:133]:
```suggestion
      description: The lookalike domain.
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+      type: String
+    - contextPath: BloxOneTD.LookalikeDomain.lookalike_host
+      description: 'The lookalike host.'
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml:136]:
```suggestion
      description: The lookalike host.
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+      type: String
+    - contextPath: BloxOneTD.LookalikeDomain.reason
+      description: 'The reason of the detection.'
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml:139]:
```suggestion
      description: The reason for the detection.
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+      type: String
+    - contextPath: BloxOneTD.LookalikeDomain.target_domain
+      description: 'The domain that was targeted by the lookalike domain.'
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense.yml:142]:
```suggestion
      description: The domain that was targeted by the lookalike domain.
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense_description.md [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+
+1. From the SSO Portal, click **User Access** from the left navigation panel; and then click the **Service API Keys** tab.
+
+2. On the Service API Keys tab, click **Create**.
+
+3. In the Create Service API Keys dialog, complete the following:
+
+  - **Name**: Enter the name of the user API key. Use a name that can identify the purpose of the key.
+
+  - **Service User**: Enter the name of the service users or choose one from the drop-down list. You can associated up to 10 service API keys per service user.
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefense/InfobloxBloxOneThreatDefense_description.md:17]:
```suggestion
  - **Service User**: Enter the name of the service users or choose one from the drop-down list. You can associate up to 10 service API keys per service user.
```

**Comment by ShirleyDenkberg** [comment]
@esharf You need to create the integration readme and the pack readme.  Let me know when these are done and I will review them.

**Review by DeanArbel — APPROVED** [review]
Great job, just a couple of notes

**Pull request merged**